compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -621,24 +655,58 @@ impl Context {
                                     };
 
                                     self.define_result(dfg, instruction, value);
-                                    return Ok(());
+                                    return Ok(true);
                                 }
                             }
                             // If there is a predicate and the index is not out of range, we can directly perform the read
                             else if index < array_size && store_value.is_none() {
                                 self.define_result(dfg, instruction, array[index].clone());
-                                return Ok(());
+                                return Ok(true);
                             }
                         }
                     }
                     AcirValue::DynamicArray(_) => (),
                 }
             }
             Type::Slice(_) => {
-                // Do nothing we only want dynamic checks here
+                // Do nothing we only want dynamic checks for slices
             }
             _ => unreachable!("ICE: expected array or slice type"),
         }
+        Ok(false)
+    }
+
+    /// We need to properly setup the inputs for array operations in ACIR.
+    /// From the original SSA values we compute the following AcirVars:
+    /// - index_var is the index of the array
+    /// - predicate_index is 0, or the index if the predicate is true
+    /// - new_value is the optional value when the operation is an array_set
+    ///      When there is a predicate, it is predicate*value + (1-predicate)*dummy, where dummy is zero.
+    ///      It is a dummy value because in the case of a false predicate, the value stored at the requested index is zeroed out
+    ///      and will never be used elsewhere in the program.
+    fn convert_array_operation_inputs(
+        &mut self,
+        dfg: &DataFlowGraph,
+        index: ValueId,
+        store_value: Option<ValueId>,
+    ) -> Result<(AcirVar, Option<AcirValue>), RuntimeError> {
+        let index_var = self.convert_numeric_value(index, dfg)?;
+        let predicate_index =
+            self.acir_context.mul_var(index_var, self.current_side_effects_enabled_var)?;
+
+        let new_value = if let Some(store) = store_value {
+            let store_value = self.convert_value(store, dfg);
+            if self.acir_context.is_constant_one(&self.current_side_effects_enabled_var) {
+                Some(store_value)
+            } else {
+                // We use zero as the dummy value. As the dummy value will never be accessed
+                // it is ok to zero out the value at the requested index
+                let zero = self.acir_context.add_constant(FieldElement::zero());
```

Comment:
No it is not OK to use zero for the dummy, because we are going to make the array set whatever is the predicate.
When the predicate is false, we want the array-set to be a no-op, and if we write 0, then it is not a no-op. 
The dummy must be the array-get value at the index

Reply:
I will update

Reply:
I went with a strategy that uses `array_get_type` using the type of the store value and the predicate index. Inside `convert_array_set_store_value` I recurse over both the `store_value` and `dummy_value` at the same time multiplying the inner store_var by the true predicate and the inner dummy_var by the false predicate. Let me know what you think.

Reply:
Hmm something looks to still be broken, need to work some more on this

Reply:
Found the fix. I moved flattening of the index to `convert_array_operation_inputs` so that it is done once rather than inside of both `array_get` and `array_set`

Reply:
sounds good